### DISCUSSION TITLE
ContextMenu component

### DISCUSSION THREAD

**Discussion by TheDarkX** [mantinedev/mantine#1049] [Feature requests]
### Usage example, including component/hook API

Mouse right click context menu 

Example:
https://mui.com/components/menus/#context-menu

**Comment by pavi2410** [comment]
I think the best and easiest way to implement this to add a new `trigger="context"` type in [Menu](https://mantine.dev/core/menu/#supported-target-elements) component

**Comment by qu1ck** [comment]
I implemented a simple context menu on top of mantine menu, if someone finds this useful:

```jsx
import { Button, Menu, MenuProps, Portal, ScrollArea } from "@mantine/core";
import React, { useCallback, useEffect, useState } from "react";

interface ContextMenuInfo {
    x: number,
    y: number,
    opened: boolean,
}

export function useContextMenu():
    [ContextMenuInfo, React.Dispatch<ContextMenuInfo>, React.MouseEventHandler<HTMLElement>] {
    const [info, setInfo] = useState<ContextMenuInfo>({ x: 0, y: 0, opened: false });

    const contextMenuHandler = useCallback<React.MouseEventHandler<HTMLElement>>((e) => {
        e.preventDefault();
        setInfo({ x: e.clientX, y: e.clientY, opened: true });
    }, [setInfo]);

    return [info, setInfo, contextMenuHandler];
}

interface ContextMenuProps extends MenuProps {
    contextMenuInfo: ContextMenuInfo,
    setContextMenuInfo: (i: ContextMenuInfo) => void,
}

export function ContextMenu({ contextMenuInfo, setContextMenuInfo, children, ...other }: ContextMenuProps) {
    const onClose = useCallback(
        () => setContextMenuInfo({ ...contextMenuInfo, opened: false }),
        [contextMenuInfo, setContextMenuInfo]);

    const [opened, setOpened] = useState<boolean>(false);

    useEffect(() => setOpened(contextMenuInfo.opened), [contextMenuInfo.opened]);

    return (
        <Menu {...other}
            opened={opened}
            onClose={onClose}
            offset={0}
            middlewares={{ shift: true, flip: false }}
            position="right-start"
        >
            <Portal>
                <Menu.Target>
                    <Button unstyled
                        sx={{
                            position: "absolute",
                            width: 0,
                            height: 0,
                            padding: 0,
                            border: 0,
                        }}
                        style={{
                            left: contextMenuInfo.x,
                            top: contextMenuInfo.y,
                        }} />
                </Menu.Target>
                <Menu.Dropdown>
                    <ScrollArea.Autosize
                        type="auto"
                        mah={`calc(${window.visualViewport!.height}px - 0.5rem)`}
                        offsetScrollbars
                        styles={{
                            viewport: {
                                paddingBottom: 0
                            }
                        }}
                    >
                        {children}
                    </ScrollArea.Autosize>
                </Menu.Dropdown>
            </Portal>
        </Menu>
    );
}
```

It works by creating a hidden button as target at the click coordinates. To use it call `useContextMenu()` hook, it will return state for ContextMenu and a mouse event handler you can pass as onContextMenu to any element.

**Comment by rmburg** [comment]
Radix UI offers a context menu component ([link](https://www.radix-ui.com/docs/primitives/components/context-menu)).
Mantine already uses Radix for other components like `ScrollArea`, maybe we could use it here too?

**Reply by rtivital** [reply]
No, Mantine uses only scroll area from radix ui, it is not planned to install other radix components.

**Comment by icflorescu** [comment]
Hey, I'm the author of [Mantine DataTable](https://icflorescu.github.io/mantine-datatable/) and I've just released the [Mantine ContextMenu](https://icflorescu.github.io/mantine-contextmenu/). It comes with a nice documentation website:

[![Mantine ContextMenu](https://user-images.githubusercontent.com/581999/231264202-4a7236c8-df59-4be8-9720-7a627c8f0334.png)](https://icflorescu.github.io/mantine-contextmenu/)

To use it, you need to wrap your application in a provider:

```tsx
import { MantineProvider } from '@mantine/core';
import { ContextMenuProvider } from 'mantine-contextmenu';

function App() {
  return (
    <MantineProvider>
      <ContextMenuProvider>
        {/* your app code here... */}
      </ContextMenuProvider>
    </MantineProvider>
  );
}
```
...then you can do something like this in your pages/components:

```tsx
import { useContextMenu } from 'mantine-contextmenu';
// other imports...

export default function GettingStartedExample() {
  const showContextMenu = useContextMenu();

  // ...

  return (
    <Picture
      image={image}
      onContextMenu={showContextMenu([
        {
          key: 'copy',
          onClick: () => { /* your click handler here */ },
        },
        {
          key: 'download',
          onClick: () => { /* your click handler here */ },
        },
      ])}
    />
  );
}
```

Please have a look at the [documentation website](https://icflorescu.github.io/mantine-contextmenu/) to see a bunch of usage examples.

And if you like it and find it useful, don't hesitate to star the repo or sponsor my work (shameless plug, I know 😊).

**Reply by yverby** [reply]
Looks good!

I noticed that when clicking on the right side of the window, if there is not enough space for `Menu`, then `Menu` moves to the right using animation from `transition` [property](https://github.com/icflorescu/mantine-contextmenu/blob/d1bfa81cfec541607481733c6a26a8fa5300f96b/package/ContextMenu.tsx#L14). It might be better to exclude `transition` for all properties by default

**Reply by icflorescu** [reply]
The transition is intentional, but since not everyone likes it, I could indeed make it optional...

**Comment by CastRome** [comment]


**Reply by icflorescu** [reply]
Hey, @CastRome, this is not the place to ask for help regarding [Mantine DataTable](https://icflorescu.github.io/mantine-datatable/).
Please consider studying the [documentation website](https://icflorescu.github.io/mantine-datatable/), browsing through (or creating a question) in the [discussions section](https://github.com/icflorescu/mantine-datatable/discussions) of the [project repository](https://github.com/icflorescu/mantine-datatable/).

Or feel free to drop me an email at the address listed in my GitHub profile if you need to hire my services, as I'm also offering paid support/consulting to developers/companies who are using my open-source projects 😉

Regarding your current issue - most likely you're not importing the styles correctly. See [here](https://github.com/icflorescu/mantine-datatable/discussions/534).

**Comment by icflorescu** [comment]
Seeing how this is still a highly requested / upvoted thread, I feel I have to reiterate my invitation to check out [Mantine ContextMenu](https://icflorescu.github.io/mantine-contextmenu/). It's well maintained, closely follows Mantine's versioning, and also comes with a nice documentation website:

[![Mantine ContextMenu](https://github.com/mantinedev/mantine/assets/581999/54213df8-48b9-4079-999c-93d78dc1bc92)](https://icflorescu.github.io/mantine-contextmenu/)
